File: packages/esm-format-go/pkg/esm/aggregate_fixtures_test.go
package esm

import (
	"path/filepath"
	"testing"
)

// TestAggregateValidFixtures asserts every tests/valid/aggregate/*.esm fixture
// parses and schema-validates cleanly through the Go loader. These exercise the
// additive aggregate/semiring schema deltas (op:"aggregate", the closed
// `semiring` enum, `ranges` { "from": <index-set> } references, and the
// model-level `index_sets` registry). Validation/round-trip only — the Go
// binding does no numeric evaluation. Cross-binding conformance bead
// ess-my4.1.5; RFC semiring-faq-unified-ir §5.1 / §5.2.
func TestAggregateValidFixtures(t *testing.T) {
	repoRoot, err := filepath.Abs(filepath.Join("..", "..", "..", ".."))
	if err != nil {
		t.Fatalf("resolve repo root: %v", err)
	}
	pattern := filepath.Join(repoRoot, "tests", "valid", "aggregate", "*.esm")
	files, err := filepath.Glob(pattern)
	if err != nil {
		t.Fatalf("glob %s: %v", pattern, err)
	}
	if len(files) == 0 {
		t.Fatalf("no .esm fixtures matched %s", pattern)
	}
	for _, path := range files {
		name := filepath.Base(path)
		t.Run(name, func(t *testing.T) {
			if _, err := Load(path); err != nil {
				t.Fatalf("expected %s to validate, got error: %v", name, err)
			}
		})
	}
}

// TestAggregateInvalidFixtures asserts every tests/invalid/aggregate/*.esm
// fixture is REJECTED. Each is a pure schema violation (unregistered semiring,
// ragged index set missing offsets/values, discrete variable missing shape,
// join not an array, join `on` pair wrong arity, refresh on a non-discrete
// variable), so Load returns a non-nil error at schema-validation time.
func TestAggregateInvalidFixtures(t *testing.T) {
	repoRoot, err := filepath.Abs(filepath.Join("..", "..", "..", ".."))
	if err != nil {
		t.Fatalf("resolve repo root: %v", err)
	}
	pattern := filepath.Join(repoRoot, "tests", "invalid", "aggregate", "*.esm")
	files, err := filepath.Glob(pattern)
	if err != nil {
		t.Fatalf("glob %s: %v", pattern, err)
	}
	if len(files) == 0 {
		t.Fatalf("no .esm fixtures matched %s", pattern)
	}
	for _, path := range files {
		name := filepath.Base(path)
		t.Run(name, func(t *testing.T) {
			if _, err := Load(path); err == nil {
				t.Fatalf("expected %s to be rejected, but it validated", name)
			}
		})
	}
}

File: packages/earthsci-toolkit/src/aggregate-fixtures.test.ts
/**
 * Schema-validation conformance tests for the additive aggregate / semiring /
 * index_sets fixtures (bead ess-my4.1.5; RFC semiring-faq-unified-ir §5.1 /
 * §5.2). Exercises the additive schema deltas — op:"aggregate", the closed
 * `semiring` enum, `ranges` { "from": <index-set> } references, and the
 * model-level `index_sets` registry — through the TypeScript binding's
 * validate + parse/serialize path. The TS binding does no numeric evaluation,
 * so this is validation / round-trip only (the numeric cross-binding
 * equivalence is asserted by the Julia / Rust / Python evaluator suites).
 */
import { describe, it, expect } from 'vitest';
import { readFileSync, readdirSync } from 'fs';
import { join } from 'path';
import { load, save, validateSchema } from './index.js';

const testsDir = join(__dirname, '../../../tests');

function esmFilesIn(dir: string): string[] {
  return readdirSync(dir)
    .filter((e) => e.endsWith('.esm'))
    .map((e) => join(dir, e))
    .sort();
}

describe('Aggregate / semiring fixtures', () => {
  describe('tests/valid/aggregate', () => {
    const validFiles = esmFilesIn(join(testsDir, 'valid', 'aggregate'));

    it('has fixtures to test', () => {
      expect(validFiles.length).toBeGreaterThan(0);
    });

    it.each(validFiles)('validates and round-trips %s', (filePath) => {
      const content = readFileSync(filePath, 'utf-8');

      // Schema-valid: the additive aggregate/semiring/index_sets fields all
      // satisfy the embedded JSON schema. We assert on the schema validator
      // directly (not the combined `validate()`): the structural pass is not
      // yet aggregate-aware — it miscounts an LHS-aggregate ODE and does not
      // recognise contracted index symbols — which is the evaluator bindings'
      // domain, tracked separately. The additive-field contract here is purely
      // schema-level.
      expect(validateSchema(JSON.parse(content))).toHaveLength(0);

      // parse -> serialize -> parse is a fixed point on the typed view.
      const original = load(content);
      const reloaded = load(save(original));
      expect(reloaded).toEqual(original);
    });
  });

  describe('tests/invalid/aggregate', () => {
    const invalidFiles = esmFilesIn(join(testsDir, 'invalid', 'aggregate'));

    it('has fixtures to test', () => {
      expect(invalidFiles.length).toBeGreaterThan(0);
    });

    it.each(invalidFiles)('rejects %s', (filePath) => {
      const content = readFileSync(filePath, 'utf-8');

      // Every invalid/aggregate fixture is a pure schema violation
      // (unregistered semiring, ragged missing offsets/values, discrete
      // missing shape, join not an array / wrong `on` arity, refresh on a
      // non-discrete variable), so the schema validator alone rejects them.
      expect(validateSchema(JSON.parse(content)).length).toBeGreaterThan(0);
    });
  });
});
